src/keygen.rs

```diff
@@ -0,0 +1,171 @@
+use crate::context::{DfnsContext, DfnsStore};
+use cggmp21::{
+    keygen::{KeygenBuilder, NonThresholdMsg},
+    security_level::SecurityLevel128,
+    supported_curves::Secp256k1,
+    ExecutionId,
+};
+use gadget_sdk::{
+    compute_sha256_hash,
+    event_listener::tangle::{
+        jobs::{services_post_processor, services_pre_processor},
+        TangleEventListener,
+    },
+    job,
+    network::round_based_compat::NetworkDeliveryWrapper,
+    tangle_subxt::tangle_testnet_runtime::api::services::events::JobCalled,
+    Error as GadgetError,
+};
+use k256::sha2::Sha256;
+use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
+use round_based::party::MpcParty;
+use sp_core::ecdsa::Public;
+use std::collections::BTreeMap;
+
+#[job(
+    id = 0,
+    params(n),
+    event_listener(
+        listener = TangleEventListener<DfnsContext, JobCalled>,
+        pre_processor = services_pre_processor,
+        post_processor = services_post_processor,
+    ),
+)]
+/// Runs a distributed key generation (DKG) process using DFNS-CGGMP21 protocol
+///
+/// # Arguments
+/// * `n` - Number of parties participating in the DKG
+/// * `context` - The DFNS context containing network and storage configuration
+///
+/// # Returns
+/// Returns the generated public key as a byte vector on success
+///
+/// # Errors
+/// Returns an error if:
+/// - Failed to retrieve blueprint ID or call ID
+/// - Failed to get party information
+/// - MPC protocol execution failed
+/// - Serialization of results failed
+pub async fn keygen(n: u16, context: DfnsContext) -> Result<Vec<u8>, GadgetError> {
+    // Get configuration and compute deterministic values
+    let blueprint_id = context
+        .blueprint_id()
+        .map_err(|e| KeygenError::ContextError(e.to_string()))?;
+    let call_id = context
+        .current_call_id()
+        .await
+        .map_err(|e| KeygenError::ContextError(e.to_string()))?;
+
+    let (meta_hash, deterministic_hash) = compute_deterministic_hashes(n, blueprint_id, call_id);
+    let execution_id = ExecutionId::new(&deterministic_hash);
+
+    // Setup party information
+    let (party_index, operators) = context
+        .get_party_index_and_operators()
+        .await
+        .map_err(|e| KeygenError::ContextError(e.to_string()))?;
+
+    let parties: BTreeMap<u16, Public> = operators
+        .into_iter()
+        .enumerate()
+        .map(|(j, (_, ecdsa))| (j as u16, ecdsa))
+        .collect();
+
+    gadget_sdk::info!(
+        "Starting DFNS-CGGMP21 Keygen for party {party_index}, n={n}, eid={}",
+        hex::encode(execution_id.as_bytes())
+    );
+
+    // Initialize RNG and network
+    let mut rng = ChaCha20Rng::from_seed(deterministic_hash);
```

Comment:
That does not look secure... `deterministic_hash` is the same across all signers (you use it as execution id), so all signers will have the same randomness.

Each signer must have reliable unpredictable source of randomness for protocol to be secure. You should use OsRng instead.